nexus/reconfigurator/execution/src/omicron_zones.rs

```diff
@@ -85,29 +100,235 @@ pub(crate) async fn deploy_zones(
     }
 }
 
+/// Idempontently perform any cleanup actions necessary for expunged zones.
+///
+/// # Panics
```

Comment:
Presumably we could instead log an error and return rather than taking out Nexus. I'm a bit concerned about taking out nexus, because it limits our debugging ability via omdb and the API. This is a pretty serious error, and hard to do accidentally though. It also leaves things in a state where invariants may be violated, which is almost certainly why a panic is the right thing to do.

 Presumably in the future we will have an E2E test that actually runs the reconfigurator and will panic if the caller in `execution/src/lib.rs` gets changed. However, even with that test we can prevent the absence of new callers doing the wrong thing. 

Thinking about whether it's better to panic or log is definitely a false dichotomy also, and only exists because we don't have other options. It would be pretty great if we could instead trigger an alert and prevent the reconfigurator from running without an override rather than killing nexus, which will end up in all Nexuses eventually panicking because of the timer based activation of the executor.

Reply:
I think we can also reasonably do _nothing_ here; just make this function filter out any non-expunged zones. It's a little less obvious, but much easier to use safely - call it with whatever zones you want, and it cleans up any expunged ones. Changed in 2d0c3882d